server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java

```diff
@@ -140,6 +140,12 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
             return in.explain(context, doc);
         }
 
+        @Override
+        public int count(LeafReaderContext context) throws IOException {
+            shardKeyMap.add(context.reader());
```

Comment:
I *think* we could not call this method since `Weight#count` would never create a cache entry in the query cache under the hood (since `count` is supposed to run in constant-time and creating a cache entry cannot possibly run in constant-time). Mentioning it in case this `ShardCoreKeyMap#add` call is responsible for the slowdown.

Reply:
Do we need to clear up whether this method call is necessary or not before we merge?

Reply:
I was raising this idea in the case when this shard key map would be the cause of the slow down that Nik had identified. Since we do not have a slowdown anymore, I have a preference for keeping this code, which is safer since it doesn't make an assumption about whether `Weight#count` might create a cache entry or not: It always registers the reader to make sure that we will associate cache entries to it when applicable, and worst-case scenario this reader never gets any cache entry associated with it and this doesn't cause any issue.